### ISSUE TITLE
Delete Junction under Windows 7

### ISSUE THREAD

**Issue by Chocobo1** [arvidn/libtorrent#331]
From issue: https://github.com/qbittorrent/qBittorrent/issues/4362

Quote from verbal666:
>THE BUG:
1. i start a magnet link download
2. qBT loads it inside his gui (0 seed 0 peer), torrent is dead.. so..
3. i delete the download inside qBT and click "also delete the files on hdd"
4. qBT deletes my junction!!!???!!! i need to redo the junction from windows disk management.

By "junction" it means it's a hard-link to a folder in Windows.

My observation:
It doesn't matter if it's hard-link or soft-link or a real dir.
The torrent state which is still trying to download metadata is related to this issue, the state which nothing is sure but the chosen download dir path.

The issue is: libtorrent deleted the "download dir" even when the torrent is a single-file torrent (by calling `remove_torrent(h, session::delete_files)`).

The delete action is triggered at libtorrent because my trace shows Session::handleTorrentRemovedAlert() is invoked (means the deletion by libtorrent is a success).

**Comment by arvidn** [comment]
pease correct me if I'm wrong, my understanding of the problem is that magnet links (before they have the metadata) have their download directory set to the directory they will be saved to, not the root of the torrent (or the file, in the case of single file torrents). When deleting the files of such torrents, the root directory and its content is deleted, because the delete logic does not realize that we just don't know what the torrent's file or directory will be yet.

I imagine that simply ignoring any attempts to delete the files of a torrent that doesn't have metadata yet would be a reasonable solution. Perhaps along with reporting the alert as if the files were successfully deleted.

Does that make sense?

**Comment by Chocobo1** [comment]
>When deleting the files of such torrents, the root directory and its content is deleted, because the delete logic does not realize that we just don't know what the torrent's file or directory will be yet.

Yes. To be sure, an example:
*single-file* torrent set to save in `/download/exp/`
`/download/exp/` is deleted in this case if it's not empty.

>I imagine that simply ignoring any attempts to delete the files of a torrent that doesn't have metadata yet would be a reasonable solution. Perhaps along with reporting the alert as if the files were successfully deleted.

Sounds good to me.

**Comment by arvidn** [comment]
attempted fix here: https://github.com/arvidn/libtorrent/pull/334

**Comment by sledgehammer999** [comment]
@Chocobo1 can you verify?

**Comment by Chocobo1** [comment]
>attempted fix here: #334

Fix confirmed.

>@Chocobo1 can you verify?

I've been busy recently.

-----

BTW, @arvidn 
There is a warning that catches my eye (compiler is gcc 5.3.0), you might be interested:

```make
CXX      session_impl.lo
session_impl.cpp: In member function ‘void libtorrent::aux::session_impl::load_state(const libtorrent::lazy_entry*)’:
session_impl.cpp:1363:19: warning: offsetof within non-standard-layout type ‘libtorrent::aux::session_impl’ is undefined [-Winvalid-offsetof]
    if (c.offset == offsetof(session_impl, m_settings))
```

**Comment by sledgehammer999** [comment]
>I've been busy recently.

No problem.
